cargo-near/src/abi.rs

```diff
@@ -27,24 +27,35 @@ pub(crate) fn generate_abi(
     crate_metadata: &CrateMetadata,
     generate_docs: bool,
 ) -> anyhow::Result<AbiRoot> {
-    fs::create_dir_all(&crate_metadata.target_directory)?;
-
-    let original_cargo_toml: toml::value::Table =
-        toml::from_slice(&fs::read(&crate_metadata.manifest_path.path)?)?;
-
-    if !original_cargo_toml
-        .get("dependencies")
-        .ok_or_else(|| anyhow::anyhow!("[dependencies] section not found"))?
-        .get("near-sdk")
-        .ok_or_else(|| anyhow::anyhow!("near-sdk dependency not found"))?
-        .as_table()
-        .ok_or_else(|| anyhow::anyhow!("near-sdk dependency should be a table"))?
-        .get("features")
-        .and_then(|features| features.as_array())
-        .map(|features| features.contains(&toml::Value::String("abi".to_string())))
-        .unwrap_or(false)
-    {
-        anyhow::bail!("Unable to generate ABI: NEAR SDK \"abi\" feature is not enabled")
+    let root_node = crate_metadata
+        .raw_metadata
+        .resolve
+        .as_ref()
+        .and_then(|dep_graph| {
+            dep_graph
+            .nodes
+            .iter()
+            .find(|node| node.id == crate_metadata.root_package.id)
+        })
+        .ok_or_else(|| anyhow::anyhow!("unable to appropriately resolve the dependency graph, perhaps your `Cargo.toml` file is malformed"))?;
+
+    let near_sdk_dep = root_node
+        .deps
+        .iter()
+        .find(|dep| dep.name == "near_sdk")
+        .and_then(|near_sdk| {
+            crate_metadata
+                .raw_metadata
+                .packages
+                .iter()
+                .find(|pkg| pkg.id == near_sdk.pkg)
+        })
+        .ok_or_else(|| anyhow::anyhow!("`near-sdk` dependency not found"))?;
+
+    for required_feature in ["__abi-generate", "__abi-embed"] {
+        if !near_sdk_dep.features.contains_key(required_feature) {
+            anyhow::bail!("unsupported `near-sdk` version. expected 4.1.* or higher");
+        }
```

Comment:
I must have missed where this conversation hapenned, but do we no longer require `abi` feature to be enabled? My only concern is this might be confusing to users that can successfully generate ABI using `cargo-near`, but who are not aware that there is an `abi` feature they can enable to get access to `NearSchema` and potentially some other stuff.

Reply:
https://github.com/near/cargo-near/pull/46

`cargo-near` can already build ABI without that requirement if all the interface types already implement `JsonSchema` or `BorshSchema`.

Ideally, we should get to a point where even `NearSchema` isn't feature-gated as well.

Reply:
Yes, I saw that, but I am asking more generally how we came to this decision of removing the `abi` feature.

> Ideally, we should get to a point where even NearSchema isn't feature-gated as well.

The problem with this is that it is unlikely to stabilize soon. `abi` is acting kind of like `unstable` in our case.

Reply:
> decision of removing the abi feature.

Just the `abi` feature _requirement_, where unnecessary.

- If you're already using features gated behind the `abi` feature, in spite of `cargo-near`, you already have the feature enabled otherwise `cargo build` doesn't pass anyway.
- And, if you're not, but all the types you do use already have `NearSchema` derived, there's no need to force that requirement.

Reply:
@itegulov do you have a reason why an ABI should not be allowed to be generated from a crate unless it uses the `abi` feature on the SDK dependency?

I see it as a benefit since a lot of devs coming in do not know how to enable features, so having the requirement, even if all of their types are already compatible, to generate the ABI seems like a little bit of a hurdle.

Reply:
Pulled this out into #50 so it's handled independently of this patch.

Reply:
No, I think it is fine to get rid of `abi`. I was just unsure of what to do with `NearSchema`. Moving under `unstable` makes sense to me.